### PR TITLE
Remarks and todos

### PR DESCRIPTION
**Pull request by maciejka**

### CODE REVIEW

File: scripts/keeper.ts [hunk 1]
```diff
@@ -27,8 +32,8 @@ function l2String(str: string): string {
   return `0x${Buffer.from(str, "utf8").toString("hex").padStart(64, "0")}`;
 }
 
-function toL1String(hex: string): string {
-  return `0x${BigInt(hex).toString(16).padEnd(64, "0")}`;
+function toL1String(l2String: BigNumber): string {
```
Comment (maciejka) [inline scripts/keeper.ts]: I would call it `cairoShortStringToByte32`. It is not obvious what it does at the first sight.

File: scripts/keeper.ts [hunk 2]
```diff
@@ -38,50 +43,73 @@ export function getRequiredEnv(key: string): string {
...
-    baseUrl = "http://localhost:8545";
-  } else if (network === "MAINNET") {
-    baseUrl =  `https://mainnet.infura.io/v3/${infuraApiKey}`;
-  } else if (network === "GOERLI") {
-    baseUrl =  `https://goerli.infura.io/v3/${infuraApiKey}`;
-  }
-  const provider = ethers.getDefaultProvider(baseUrl);
+  const { ethereum: providerUrl } = getProviderUrls(network);
+  const provider = ethers.getDefaultProvider(providerUrl);
   const mnemonic = getRequiredEnv("MNEMONIC");
```
Comment (maciejka) [inline scripts/keeper.ts]: It is more practical to user private key than a mnemonic.

File: scripts/keeper.ts [hunk 3]
```diff
@@ -171,42 +178,76 @@ async function finalizeFlush(targetDomain: string) {
...
+        l2WormholeGatewayAddress,
+        l1WormholeGatewayAddress
+      );
+      const consumedMessageEvents = (
+        await starknet.queryFilter(consumedMessageFilter, event.blockNumber)
+      ).filter((consumedEvent) => {
+        const eventArgs = event.args.map((_) => _.toString());
+        const consumedEventArgs = consumedEvent.args.map((_) => _.toString());
+        return (
+          consumedEventArgs[0] === eventArgs[0] &&
```
Comment (maciejka) [inline scripts/keeper.ts]: Why not `isEqual` from lodash?

File: scripts/keeper.ts [hunk 4]
```diff
@@ -171,42 +178,76 @@ async function finalizeFlush(targetDomain: string) {
...
+          consumedEventArgs[1] === eventArgs[1] &&
+          consumedEventArgs[2] === eventArgs[2] &&
+          consumedEventArgs[3] === eventArgs[3]
         );
       });
-      if (consumedMessageEvents.length == 0) {
+      if (consumedMessageEvents.length === 0) {
```
Comment (maciejka) [inline scripts/keeper.ts]: I would extract logic starting around line 204 into a separate function called ~ `flushesToBeFinalized`

File: scripts/keeper.ts [hunk 5]
```diff
@@ -93,76 +121,55 @@ function getL2Signer(network: string): starknet.Account {
...
 
+  const wormholeJoinAddress = getRequiredEnv(
+    `${NETWORK}_WORMHOLE_JOIN_ADDRESS`
+  );
+  const wormholeJoin = await getL1ContractAt<WormholeJoin>(
+    l1Signer,
+    "WormholeJoin",
+    wormholeJoinAddress
+  );
+  const settleFilter = wormholeJoin.filters.Settle(l1String(SOURCE_DOMAIN));
```
Comment (maciejka) [inline scripts/keeper.ts]: Extract lines ~ 154 - 165 into a function. Let's call it recentFlushTimestamp. FLUSH_DELAY should be in seconds.

File: scripts/keeper.ts [hunk 6]
```diff
@@ -38,50 +43,73 @@ export function getRequiredEnv(key: string): string {
   return value;
 }
 
-export async function getL1ContractAt(signer: any, name: string, address: string) {
+function getProviderUrls(network: string): {
```
Comment (maciejka) [inline scripts/keeper.ts]: This should be kept in a .env file. With dotenv-expand probably. E.g.:
INFURA_API_KEY=...
ALPHA_MAINNET_ETHEREUM_PROVIDER_URL=...${INFURA_API_KEY}
ALPHA_MAINNET_STARKNET_PROVIDER_URL=...${INFURA_API_KEY}

File: scripts/keeper.ts [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
+  L1DAIWormholeGateway,
+  Starknet,
+  WormholeJoin,
+} from "types/ethers-contracts";
 import { l2_dai_wormhole_gateway } from "types/starknet-contracts/l2_dai_wormhole_gateway";
 dotenv.config();
 
 const NETWORK = getRequiredEnv("NETWORK").toUpperCase();
 const ENV = getRequiredEnv("NODE_ENV").toUpperCase();
+const START_BLOCK = parseInt(getRequiredEnv("START_BLOCK"));
```
Comment (maciejka) [inline scripts/keeper.ts]: It seems like sth that will need to be updated. Why not try to look back not futher than FLUSH_DELAY? Aproximate timestamp to block is here: https://github.com/makerdao/xdomain/blob/master/packages/wormhole-keeper/src/utils.ts

Comment (nulven) [reply scripts/keeper.ts]: This is only to keep from looking through blocks way before the contracts are even deployed, it was causing errors

Comment (maciejka) [reply scripts/keeper.ts]: My point is that we can have one parameter less in the configuration for relatively small price (by using findNearestBlock).

File: scripts/utils.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+export function getConfig() {
+  const localEnv = dotenv.config({ path: path.resolve(process.cwd(), ".env.local") });
+  dotenvExpand.expand(localEnv);
+  if (process.env.NODE_ENV === "dev") {
+    const forkEnv = dotenv.config({ path: path.resolve(process.cwd(), ".env.fork") });
+    dotenvExpand.expand(forkEnv);
+  }
+  const env = dotenv.config({ path: path.resolve(process.cwd(), ".env") });
+  dotenvExpand.expand(env);
```
Comment (nulven) [inline scripts/utils.ts:46]: Whichever gets loaded first does not get overwritten

File: scripts/keeper.ts [hunk 9]
```diff
@@ -1,217 +1,162 @@
...
-    baseUrl = "https://alpha4.starknet.io";
+  const encodedDomain = l2String(targetDomain);
+  async function recentFlushTimestamp(): Promise<number> {
+    const wormholeJoin = await getL1ContractAt<WormholeJoin>(
+      l1Signer,
+      "WormholeJoin",
+      config.wormholeJoinAddress
+    );
+    const settleFilter = wormholeJoin.filters.Settle(l1String(config.sourceDomain));
+    const nearestBlock = await findNearestBlock(l1Signer.provider, Date.now());
```
Comment (nulven) [inline scripts/keeper.ts]: Not sure what the timestamp should be

Comment (maciejka) [reply scripts/keeper.ts]: `now - 10 * flush_delay` ?

File: scripts/keeper.ts [hunk 10]
```diff
@@ -1,221 +1,159 @@
...
+  l2String,
+  toUint,
+  getConfig,
+  findNearestBlock,
+  Config,
+} from "./utils";
 
-const SOURCE_DOMAIN = `${NETWORK}-SLAVE-STARKNET-1`;
-const FLUSH_DELAY = 100;
+async function flush(targetDomain: string) {
```
Comment (maciejka) [inline scripts/keeper.ts]: I would rather make target domains a config parameter. It should work for multiple domains.

File: scripts/keeper.ts [hunk 11]
```diff
@@ -1,221 +1,159 @@
...
+  );
 
-function l2String(str: string): string {
-  return `0x${Buffer.from(str, "utf8").toString("hex").padStart(64, "0")}`;
-}
+  const encodedTargetDomain = l2String(targetDomain);
 
-function toL1String(hex: string): string {
-  return `0x${BigInt(hex).toString(16).padEnd(64, "0")}`;
+  const lastFlushTimestamp = await recentFlushTimestamp(config, l1Signer);
```
Comment (maciejka) [inline scripts/keeper.ts]: `recentFlushTimestamp` should either return pairs (`lastTimestamp`, `targetDomain`) or take `targetDomain` as parameter.

File: scripts/keeper.ts [hunk 12]
```diff
@@ -1,221 +1,159 @@
...
+  const [daiToFlushSplit] = await l2WormholeGateway.batched_dai_to_flush(
+    encodedTargetDomain
+  );
+  const daiToFlush = toUint(daiToFlushSplit);
+  console.log(`DAI to flush: ${daiToFlush}`);
+
+  if (daiToFlush > 0 && Date.now() > lastFlushTimestamp + config.flushDelay) {
```
Comment (maciejka) [inline scripts/keeper.ts]: I would rather:
```
daiToFlush > config.flushMinimum  || daiToFlush > 0 && Date.now() > lastFlushTimestamp + config.flushDelay
```
It does not make sense to flush pennies.

File: .env.fork [hunk 13]
```diff
@@ -0,0 +1,20 @@
+NETWORK=alpha-goerli
```
Comment (maciejka) [inline .env.fork]: Shouldn't `.env.fork` include just `*_PROVIDER` keys?

File: scripts/keeper.ts [hunk 14]
```diff
@@ -1,221 +1,163 @@
...
-  const accountContract = new starknet.Contract(
-    compiledArgentAccount.abi,
-    accountResponse.address,
-    provider
+  const settleFilter = wormholeJoin.filters.Settle(l1String(sourceDomain));
+  const nearestBlock = await findNearestBlock(l1Signer.provider, Date.now() - 10 * flushDelay);
+  const settleEvents = await wormholeJoin.queryFilter(
+    settleFilter,
+    6800000,
+    // nearestBlock
```
Comment (maciejka) [inline scripts/keeper.ts]: Commented for testing?

File: scripts/keeper.ts [hunk 15]
```diff
@@ -45,9 +48,13 @@ export async function flush(config: Config) {
     Date.now() > lastFlushTimestamp + config.flushDelay
   ) {
     console.log("Sending `flush` transaction");
-    const { transaction_hash } = await l2WormholeGateway.flush(
+    const { amount } = await l2TeleportGateway.estimate(
```
Comment (maciejka) [inline scripts/keeper.ts]: There is an estimateFee version of this method generated by TypeChain:
l2TeleportGateway.estimateFee.flush

File: scripts/keeper.ts [hunk 16]
```diff
@@ -16,25 +16,28 @@ import {
   getL1Signer,
   getL2ContractAt,
   getL2Signer,
+  getRequiredEnv,
   l1String,
   l2String,
   toUint,
 } from "./utils";
 
+const FEE_MULTIPLIER = parseInt(getRequiredEnv("FEE_MULTIPLIER"));
```
Comment (maciejka) [inline scripts/keeper.ts]: Is it l1 or l2 fee?

**Review by maciejka — COMMENTED** [review]
In general TypeChain docs recommends not to commit generated files to the vcs.

File: scripts/keeper.ts (hunk 11 above)
Comment (nulven) [reply scripts/keeper.ts]: Not sure what you're thinking. The targetDomain would determine the address of the TeleportJoin contract that we query

Comment (maciejka) [reply scripts/keeper.ts]: Ideally this function should observe `Flushed` events on `l2_dai_teleport_gateway` and return timestamps of last flush per target domain. Until observing events is available we use `Settle` events on `TeleportJoin` which is less general. Nonetheless function should return pairs (timestamp, targetDomain). Source and target domains should not be in the config file. These are redundant with regard to teleport gateways and temporarily teleportJoin.

File: scripts/keeper.ts [hunk 17]
```diff
@@ -1,221 +1,171 @@
...
-async function finalizeFlush(targetDomain: string) {
-  const l1Signer = getL1Signer(NETWORK);
-  const l2Signer = getL2Signer(NETWORK);
-
-  const l1WormholeGatewayAddress = getRequiredEnv(`${NETWORK}_L1_DAI_WORMHOLE_GATEWAY_ADDRESS`);
-  const l2WormholeGatewayAddress = getRequiredEnv(`${NETWORK}_L2_DAI_WORMHOLE_GATEWAY_ADDRESS`);
-  const starknetAddress = getRequiredEnv(`${NETWORK}_STARKNET_ADDRESS`);
-  const l1WormholeGateway: L1DAIWormholeGateway = (await getL1ContractAt(l1Signer, "L1DAIWormholeGateway", l1WormholeGatewayAddress) as L1DAIWormholeGateway);
-  const l2WormholeGateway: l2_dai_wormhole_gateway = (await getL2ContractAt(l2Signer, "l2_dai_wormhole_gateway", l2WormholeGatewayAddress) as l2_dai_wormhole_gateway);
+async function flushesToBeFinalized(
```
Comment (maciejka) [inline scripts/keeper.ts:101]: Can't we simply query `LogMessageToL1` with `fromBlock` set to either: `Date.now() - 10 * config.flushDelay`, or if there are no results preconfigured `startBlock`?
Btw: I would make `flushDelay multiplier` and `startingBlock` a .env parameters.

File: scripts/keeper.ts [hunk 18]
```diff
@@ -1,221 +1,171 @@
...
+  l1Signer: Signer,
+  starknet: Starknet,
+  config: ReturnType<typeof getConfig>
+): Promise<Event[]> {
```
Comment (maciejka) [inline scripts/keeper.ts]: Return type should be: `Promise<LogMessageToL1Event[]>`?

File: scripts/keeper.ts [hunk 19]
```diff
@@ -1,221 +1,171 @@
...
-        console.log("Sending `finalizeFlush` transaction");
-	const tx = await l1WormholeGateway.finalizeFlush(toL1String(event.args.payload[1]), event.args.payload[2]);
-	const res = await tx.wait();
-	console.log("Success");
-	fs.writeFileSync("lastFlushBlock", event.blockNumber.toString());
-      } else {
-	console.log("Flush message was consumed");
-      }
+  async function getConsumedSettleMessageEvents(
+    event: Event
```
Comment (maciejka) [inline scripts/keeper.ts]: Return type should be: Promise<ConsumedLogMessageToL1Event[]>?

File: scripts/keeper.ts [hunk 20]
```diff
@@ -1,221 +1,171 @@
...
-}
 
-if (process.argv[2] === "flush") {
-  flush(process.argv[3]);
-} else if (process.argv[2] === "finalizeFlush") {
-  finalizeFlush(process.argv[3]);
+  const logMessageEvents = await getSettleMessageEvents();
+  return Promise.all(
+    logMessageEvents.filter(async (event: Event) => {
+      const consumedMessageEvents = await getConsumedSettleMessageEvents(event);
```
Comment (maciejka) [inline scripts/keeper.ts:141]: `payload[0] === HANDLE_FLUSH` should be checked here

File: scripts/keeper.ts [hunk 21]
```diff
@@ -1,221 +1,171 @@
...
   }
-}
 
-if (process.argv[2] === "flush") {
-  flush(process.argv[3]);
-} else if (process.argv[2] === "finalizeFlush") {
-  finalizeFlush(process.argv[3]);
+  const logMessageEvents = await getSettleMessageEvents();
+  return Promise.all(
+    logMessageEvents.filter(async (event: Event) => {
```
Comment (maciejka) [inline scripts/keeper.ts]: event type should be: `LogMessageToL1Event`

File: scripts/keeper.ts [hunk 22]
```diff
@@ -1,156 +1,127 @@
...
-    encodedTargetDomain
+  const logMessageFilter = starknet.filters.LogMessageToL1(
+    l2TeleportGateway.address,
+    l1TeleportGatewayAddress
   );
-  const daiToFlush = toUint(daiToFlushSplit);
-  console.log(`DAI to flush: ${daiToFlush}`);
+  const logMessageEvents = await starknet.queryFilter(
+    logMessageFilter,
+    findNearestBlock(l1Signer.provider, Date.now() - flushDelayMultiplier * flushDelay)
```
Comment (maciejka) [inline scripts/keeper.ts]: `await` seems to be missing here?

File: scripts/keeper.ts [hunk 23]
```diff
@@ -1,156 +1,127 @@
...
-      encodedTargetDomain,
-      { maxFee: amount * config.l2GasMultiplier }
-    );
-    await l2Signer.waitForTransaction(transaction_hash);
-    console.log("Success");
-  } else {
-    console.log("Flush not needed");
-  }
+  // eventually get targetDomains from File events
+  const targetDomains = ["GOERLI-MASTER-1"];
```
Comment (maciejka) [inline scripts/keeper.ts]: Let's put it into the configuration until events are available on l2.

File: scripts/keeper.ts [hunk 24]
```diff
@@ -1,156 +1,127 @@
...
+  const domainsToFlush = [];
+
+  flushEvents.forEach(async (event) => {
+    const targetDomain = cairoShortStringToBytes32(BigNumber.from(event.args[1]));
+    domainsToFlush.push(targetDomain);
+  });
+  return domainsToFlush;
```
Comment (maciejka) [inline scripts/keeper.ts:60]: Domain and timestamp of recent timestamp should be returned here.

File: scripts/keeper.ts [hunk 25]
```diff
@@ -1,156 +1,127 @@
...
-      { maxFee: amount * config.l2GasMultiplier }
-    );
-    await l2Signer.waitForTransaction(transaction_hash);
-    console.log("Success");
-  } else {
-    console.log("Flush not needed");
-  }
+  // eventually get targetDomains from File events
+  const targetDomains = ["GOERLI-MASTER-1"];
+  const domainsToFlush = [];
```
Comment (maciejka) [inline scripts/keeper.ts:54]: Shouldn't `targetDomains` be used to initialize domainsToFlush? I mean to sth like:
```
{ 
  'GOERLI-MASTER-1': 0 // 0 is previous flush timestamp, should be updated by flushEvents query 
}
```
?

File: scripts/keeper.ts [hunk 26]
```diff
@@ -1,156 +1,127 @@
...
-    nearestBlock
-  );
-  const recentEvent = settleEvents[settleEvents.length - 1];
-  if (recentEvent) {
-    const block = await recentEvent.getBlock();
-    return block.timestamp;
-  } else {
-    return 0;
-  }
+    if (daiToFlush > config.flushMinimum) {
```
Comment (maciejka) [inline scripts/keeper.ts:81]: Why flushDelay is ignored here?

File: scripts/keeper.ts [hunk 27]
```diff
@@ -1,156 +1,127 @@
...
+      console.log("Sending `flush` transaction");
```
Comment (maciejka) [inline scripts/keeper.ts]: In general logs should be more meaningful, in this case `targetDomain` and, `daiToFlush` should be included.

File: scripts/keeper.ts [hunk 28]
```diff
@@ -161,11 +132,45 @@ async function flushesToBeFinalized(
...
+  const l1TeleportGateway = await getL1ContractAt<L1DAITeleportGateway>(
+    l1Signer,
+    "L1DAITeleportGateway",
+    l1TeleportGatewayAddress
+  );
+
+  const flushes = await flushesToBeFinalized(l1TeleportGatewayAddress, starknet, l1Signer, config);
+  for (let i = 1; i < flushes.length; i++) {
+    const flush = flushes[i];
+    console.log("Sending `finalizeFlush` transaction");
```
Comment (maciejka) [inline scripts/keeper.ts]: Include targetDomain and amount in log message.

File: scripts/keeper.ts [hunk 29]
```diff
@@ -161,11 +132,45 @@ async function flushesToBeFinalized(
...
+    const tx = await l1TeleportGateway.finalizeFlush(
+      cairoShortStringToBytes32(flush.args.payload[1]),
+      flush.args.payload[2]
```
Comment (maciejka) [inline scripts/keeper.ts:175]: Add gasPrice management.

File: scripts/keeper.ts [hunk 30]
```diff
@@ -1,156 +1,127 @@
...
-  const l2TeleportGateway = await getL2ContractAt<l2_dai_teleport_gateway>(
-    l2Signer,
-    "l2_dai_teleport_gateway",
-    config.l2TeleportGatewayAddress
+async function getDomainsToFlush(
+  l2TeleportGateway: l2_dai_teleport_gateway,
+  l1Signer: Signer,
+  { flushDelay, flushDelayMultiplier, starknetAddress }: Config
+): Promise<string[]> {
+  const l1TeleportGatewayAddress = `0x${(await l2TeleportGateway.teleport_gateway())[0].toString(16)}`;
```
Comment (maciejka) [inline scripts/keeper.ts]: Use `getL1TeleportGatewayAddress` here.

File: scripts/keeper.ts [hunk 31]
```diff
@@ -1,221 +1,176 @@
...
-  return contractFactory.attach(address);
+  const domainsToFlush = await getDomainsToFlush(l2TeleportGateway, l1Signer, config);
+  domainsToFlush.forEach(async (targetDomain) => {
+    const [daiToFlushSplit] = await l2TeleportGateway.batched_dai_to_flush(
+      targetDomain
+    );
+    const daiToFlush = toUint(daiToFlushSplit);
+    console.log(`DAI to flush: ${daiToFlush}`);
+
+    if (daiToFlush > config.flushMinimum) {
```
Comment (maciejka) [inline scripts/keeper.ts]: Per domain we should:
1. check if `daiToFlush > config.flushMinimum`?
2. if yes `flush`
3. if not and `daiToFlush > 0`, were there any flushes after `now - flushDelay`?
4. if yes `flush`

File: scripts/keeper.ts (hunk 29 above)
Comment (nulven) [reply scripts/keeper.ts:175]: gas price management is only needed on l2 calls

Comment (maciejka) [reply scripts/keeper.ts:175]: Usually one wants to be able to mange gas price on l1. For one time scripts overriding gasPrice is usually enough. For an autonomous keeper strategy should be more elaborate. I am not sure how to do it in context of EIP-1559. Let's turn this problem into a separate task.

File: scripts/keeper.ts [hunk 32]
```diff
@@ -71,24 +70,30 @@ export async function flush(config: Config) {
...
         targetDomain,
         { maxFee: amount * config.l2GasMultiplier }
       );
       await l2Signer.waitForTransaction(transaction_hash);
       console.log("Success");
+    }
+
+    if (daiToFlush > config.flushMinimum) {
+      await sendFlush();
+    } else if (daiToFlush > 0 && domainsToFlush.includes(targetDomain)) {
```
Comment (maciejka) [inline scripts/keeper.ts]: it should be: `daiToFlush > 0 && !domainsToFlush.includes(targetDomain)` ? I mean flush if there were no recent flushes.

File: scripts/keeper.ts [hunk 33]
```diff
@@ -71,24 +70,30 @@ export async function flush(config: Config) {
...
     const daiToFlush = toUint(daiToFlushSplit);
     console.log(`DAI to flush: ${daiToFlush}`);
 
-    if (daiToFlush > config.flushMinimum) {
+    const sendFlush = async () => {
       console.log("Sending `flush` transaction");
-      const { amount } = await l2TeleportGateway.estimateFee.flush([
-        targetDomain 
-      ]);
+      console.log(` Domain: ${targetDomain}`);
```
Comment (maciejka) [inline scripts/keeper.ts]: Log messages should be one liners.

File: scripts/keeper.ts [hunk 34]
```diff
@@ -43,14 +43,13 @@ async function getDomainsToFlush(
   );
   const logMessageEvents = await starknet.queryFilter(
     logMessageFilter,
-    findNearestBlock(l1Signer.provider, Date.now() - flushDelayMultiplier * flushDelay)
+    await findNearestBlock(l1Signer.provider, Date.now() - flushDelayMultiplier * flushDelay)
```
Comment (maciejka) [inline scripts/keeper.ts]: We want recent flushes here. `flushDelayMultiplier` should be ~ 1. But `flushesToBeFinalized`  expects `flushesToBeFinalized` to be higher.

Comment (nulven) [reply scripts/keeper.ts:46]: I don't think we need flushDelayMultiplier anymore

**Pull request merged**